File: 2667_단지번호붙이기.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int n, cnt;
int town[25][25];
int visited[25][25];
vector<int> houseCnt;
int dx[] = {-1, 0, 1, 0};
int dy[] = {0, 1, 0, -1};

int dfs(int x, int y) {
    // 인덱스가 town을 넘어가는 경우 제외
    if (x < 0 || y < 0 || x >= n || y >= n) return 0;
    // 집이 있고 처음 방문하는 경우에만 실행
    if (town[x][y] != 0 && visited[x][y] == 0) {
        // 방문 체크
        visited[x][y] = 1;
        // 단지에 포함되는 집 개수
        cnt += 1;
        // 상하좌우 집이 있는가 판별
        for (int i = 0; i < 4; i++) {
            dfs(x + dx[i], y + dy[i]);
        }
        return 1;
    }
    return 0;
}

int main() {
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    
    cin >> n;
    for (int i = 0; i < n; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < s.length(); j++) {
            town[i][j] = s[j] - '0';
        }
    }
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            // 모든 집 다 방문
            cnt = 0;
            // 단지 하나
            int tmp = dfs(i, j);
            // 단지가 생성됬다면 집 개수 push
            if (tmp != 0) {
                houseCnt.push_back(cnt);
            }
        }
    }
    
    cout << houseCnt.size() << endl;
    sort(houseCnt.begin(), houseCnt.end());
    for(int i = 0; i < houseCnt.size(); i++) {
        cout << houseCnt[i] << endl;
    }
}

File: 2667_단지번호붙이기.swift
import Foundation

let n = Int(readLine()!)!
var town = [[Int]](repeating: [Int](repeating: 0, count: n), count: n)
var visited = [[Bool]](repeating: [Bool](repeating: false, count: n), count: n)
var houseCnt = [Int]()
var cnt = 0
let dx = [-1, 0, 1, 0]
let dy = [0, 1, 0, -1]

for i in 0..<n {
    let s = readLine()!.map{Int(String($0))!}
    for j in 0..<n {
        town[i][j] = s[j]
    }
}

for i in 0..<n {
    for j in 0..<n {
        cnt = 0
        dfs(x: i, y: j)
        if cnt != 0 {
            houseCnt.append(cnt)
        }
    }
}

print(houseCnt.count)
houseCnt.sorted().forEach {
    print($0)
}


func dfs(x: Int, y: Int) {
    if x<0 || y<0 || x>=n || y>=n { return }
    if town[x][y] != 0 && !visited[x][y] {
        visited[x][y] = true
        cnt += 1
        for i in 0..<4{
            dfs(x: x + dx[i], y: y + dy[i])
        }
    }
}
